File: AnylineExamples/Anyline Examples Source/Utility/View/ResultScreen/ALResultViewController.m
//
//  ALResultViewController.m
//  AnylineExamples
//
//  Created by Philipp Müller on 07.05.18.
//

#import "ALResultViewController.h"
#import <Foundation/Foundation.h>
#import "UIFont+ALExamplesAdditions.h"
#import "ALResultCell.h"
#import "ALResultEntry.h"
#import "UIColor+ALExamplesAdditions.h"
#import "NSAttributedString+ALExamplesAdditions.h"
#import "ALUniversalIDFieldnameUtil.h"

static NSString *disclaimerString = @"The result fields above display a selection of scannable ID information only. Please review the documentation for a full list of scannable ID information.";
static NSString *WEBLINK_ANYLINE_DOCUMENTATION_PRODUCTID = @"https://documentation.anyline.com/toc/products/id/index.html";

@interface ALResultViewController () <UITableViewDelegate, UITableViewDataSource, UIScrollViewDelegate>

@property (strong, nonatomic) NSArray *entryArray;
@property (strong, nonatomic) IBOutlet UIScrollView *contentScrollView;
@property (strong, nonatomic) IBOutlet UIImageView *faceImageView;
@property (strong, nonatomic) IBOutlet UITableView *tableView;
@property (strong, nonatomic) IBOutlet UIImageView *firstImageView;
@property (strong, nonatomic) IBOutlet UIImageView *secondImageView;
@property (strong, nonatomic) IBOutlet UITextView *disclaimerTextView;
@property (strong, nonatomic) IBOutlet UIButton *confirmButton;

@property (nonatomic, strong) NSLayoutConstraint *tableViewHeightConstraint;

@end


@implementation ALResultViewController

- (instancetype)initWithResultData:(NSMutableArray<ALResultEntry *>*)resultData image:(UIImage *)image {
    return [self initWithResultDataDictionary:@{ @"Result Data" : resultData}
                                        image:image
                                optionalImage:nil
                                    faceImage:nil
                         shouldShowDisclaimer:NO];
}

- (instancetype)initWithResultData:(NSMutableArray<ALResultEntry *>*)resultData
                             image:(UIImage *)image
              shouldShowDisclaimer:(BOOL)shouldShow {
    return [self initWithResultDataDictionary:@{ @"Result Data" : resultData}
                                        image:image
                                optionalImage:nil
                                    faceImage:nil
                         shouldShowDisclaimer:shouldShow];
}

- (instancetype)initWithResultData:(NSMutableArray<ALResultEntry *>*)resultData
                             image:(UIImage *)image
                         faceImage:(UIImage *)faceImage
              shouldShowDisclaimer:(BOOL)shouldShow {
    return [self initWithResultDataDictionary:@{ @"Result Data" : resultData}
                                        image:image
                                optionalImage:nil
                                    faceImage:faceImage
                         shouldShowDisclaimer:shouldShow];
}

- (instancetype)initWithResultData:(NSMutableArray<ALResultEntry *>*)resultData
                             image:(UIImage *)image
                     optionalImage:(UIImage *)optImage
              shouldShowDisclaimer:(BOOL)shouldShow {
    return [self initWithResultDataDictionary:@{ @"Result Data" : resultData}
                                        image:image
                                optionalImage:optImage
                                    faceImage:nil
                         shouldShowDisclaimer:shouldShow];
}

- (instancetype)initWithResultData:(NSMutableArray<ALResultEntry *>*)resultData
                             image:(UIImage *)image
                     optionalImage:(UIImage *)optImage
                         faceImage:(UIImage *)faceImage
              shouldShowDisclaimer:(BOOL)shouldShow {
    return [self initWithResultDataDictionary:@{ @"Result Data" : resultData}
                                        image:image
                                optionalImage:optImage
                                    faceImage:faceImage
                         shouldShowDisclaimer:shouldShow];
}

- (instancetype)initWithResultDataDictionary:(NSDictionary *)resultDataDictionary
                                       image:(UIImage *)image
                               optionalImage:(UIImage *)optImage
                                   faceImage:(UIImage *)faceImage {
    return [self initWithResultDataDictionary:resultDataDictionary image:image optionalImage:optImage faceImage:faceImage shouldShowDisclaimer:NO];;
}


- (instancetype)initWithResultDataDictionary:(NSDictionary *)resultDataDictionary
                                       image:(UIImage *)image
                               optionalImage:(UIImage *)optImage
                                   faceImage:(UIImage *)faceImage
                        shouldShowDisclaimer:(BOOL)shouldShow {
    self = [super init];
    if (self) {
        _documentImage = image;
        _resultData = resultDataDictionary;
        _documentBackImage = optImage;
        _faceImage = faceImage;
        _shouldShowDisclaimer = shouldShow;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.leftItemsSupplementBackButton = YES;
    self.navigationItem.title = @"Result Data";
    UIBarButtonItem *shareBarItem;
    
    shareBarItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAction target:self action:@selector(askToShareItems:)];
    [shareBarItem setTintColor:[UIColor AL_LabelBlackWhite]];

    
    self.navigationItem.rightBarButtonItem = shareBarItem;
    self.view.backgroundColor = [UIColor AL_BackgroundColor];
    [self setupScrollView];
    [self setupScrollViewContent];
    [self setupConfirmButton];
    [self setupContraints];
}

- (void)askToShareItems:(id)sender {
    NSArray *resultDataArray = [self.resultData objectForKey:@"Result Data"];
    NSMutableString *shareText = [NSMutableString stringWithString:@"Anyline Scanner Result\n\n"];
    
    for (ALResultEntry *entry in resultDataArray) {
        [shareText appendFormat:@"%@: %@\n", entry.title, entry.value];
    }
    
    UIActivityViewController *activityController =  [[UIActivityViewController alloc] initWithActivityItems:@[shareText, self.documentImage]
                                                                                      applicationActivities:nil];
    [activityController setValue:@"Anyline Scanner Result" forKey:@"Subject"];
    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone) {
        [self presentViewController:activityController animated:YES completion:nil];
    } else {
        // Change Rect to position Popover
        UIPopoverController *popup = [[UIPopoverController alloc] initWithContentViewController:activityController];
        [popup presentPopoverFromRect:CGRectMake(self.view.frame.size.width/2, self.view.frame.size.height/4, 0, 0)inView:self.view permittedArrowDirections:UIPopoverArrowDirectionAny animated:YES];
    }
}

- (void)setupScrollView {
    //Setup Content ScrollView
    self.contentScrollView = [[UIScrollView alloc] init];
    self.contentScrollView.delegate = self;
    self.contentScrollView.showsVerticalScrollIndicator = YES;
    self.contentScrollView.scrollEnabled = YES;
    self.contentScrollView.userInteractionEnabled = YES;
    self.contentScrollView.backgroundColor = [UIColor AL_BackgroundColor];
    [self.view addSubview:self.contentScrollView];
    [self.contentScrollView setTranslatesAutoresizingMaskIntoConstraints:NO];
}

- (void)setupConfirmButton {
    //Setup Confirm Button
    self.confirmButton = [[UIButton alloc] init];
    [self.confirmButton addTarget:self action:@selector(confirmAction:) forControlEvents:UIControlEventTouchUpInside];
    [self.confirmButton setTitle:@"Confirm" forState:UIControlStateNormal];
    [self.confirmButton.titleLabel setFont:[UIFont AL_proximaBoldWithSize:18]];
    [self.confirmButton.titleLabel setTextColor:[UIColor whiteColor]];
    self.confirmButton.backgroundColor = [UIColor AL_examplesBlue];
    [self.confirmButton.layer setCornerRadius:50/2];
    [self.view addSubview:self.confirmButton];
    [self.confirmButton setTranslatesAutoresizingMaskIntoConstraints:NO];
}

- (void)setupScrollViewContent {
    
    self.faceImageView = [[UIImageView alloc] init];
    self.faceImageView.image = self.faceImage;
    
    [self.contentScrollView addSubview:self.faceImageView];
    [self.faceImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    //Setup TableView
    self.tableView = [[UITableView alloc] init];
    [self.tableView registerClass:[ALResultCell class] forCellReuseIdentifier:@"alResultCell"];
    [self.tableView setSectionHeaderHeight:0];
    self.tableView.userInteractionEnabled = true;
    self.tableView.scrollEnabled = NO;
    self.tableView.allowsSelection = NO;
    self.tableView.backgroundColor = [UIColor AL_BackgroundColor];
    [self.tableView setSeparatorStyle:UITableViewCellSeparatorStyleNone];
    self.tableView.estimatedRowHeight = 60;
    self.tableView.estimatedSectionFooterHeight = 0;
    self.tableView.estimatedSectionHeaderHeight = 0;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.tableFooterView = [[UIView alloc] init];
    
    //Add view to scrollView
    [self.contentScrollView addSubview:self.tableView];
    [self.tableView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    //Setup Image View
    self.firstImageView = [[UIImageView alloc] init];
    self.firstImageView.image = self.documentImage;
    self.firstImageView.contentMode = UIViewContentModeScaleAspectFit;
    [self.contentScrollView addSubview:self.firstImageView];
    [self.firstImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    //Setup Image View
    self.secondImageView = [[UIImageView alloc] init];
    self.secondImageView.image = self.documentBackImage;
    self.secondImageView.contentMode = UIViewContentModeScaleAspectFit;
    [self.contentScrollView addSubview:self.secondImageView];
    [self.secondImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    if (_shouldShowDisclaimer) {
        self.disclaimerTextView = [[UITextView alloc] init];
        [self setHyperlinkInsideTextView];
        [self.disclaimerTextView setScrollEnabled:NO];
        [self.disclaimerTextView setClipsToBounds:NO];
        [self.disclaimerTextView setUserInteractionEnabled:YES];
        [self.disclaimerTextView setEditable:NO];
        [self.contentScrollView addSubview:self.disclaimerTextView];
        [self.disclaimerTextView setTranslatesAutoresizingMaskIntoConstraints:NO];
    }
}

- (void)setupContraints {
    CGFloat verticalPadding = 10;
    CGFloat horizontalPadding = 30;
    NSMutableArray *constraints = @[[self.contentScrollView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
                                    [self.contentScrollView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
                                    [self.contentScrollView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor]].mutableCopy;
    NSLayoutYAxisAnchor *topTableVieWConstraints = self.contentScrollView.topAnchor;
    if (self.faceImage) {
        NSLayoutConstraint *horizontalConstraint;
        if (self.isArabicScript) {
            horizontalConstraint = [self.faceImageView.trailingAnchor constraintEqualToAnchor:self.contentScrollView.trailingAnchor constant:-horizontalPadding];
        } else {
            horizontalConstraint = [self.faceImageView.leadingAnchor constraintEqualToAnchor:self.contentScrollView.leadingAnchor constant:horizontalPadding];
        }
        [constraints addObjectsFromArray:@[[self.faceImageView.topAnchor constraintEqualToAnchor:self.contentScrollView.topAnchor constant:verticalPadding],
                                           horizontalConstraint,
                                           [self.faceImageView.widthAnchor constraintLessThanOrEqualToConstant:100],
                                           [self.faceImageView.heightAnchor constraintLessThanOrEqualToConstant:100]]];
        topTableVieWConstraints = self.faceImageView.bottomAnchor;
    }
    
    self.tableViewHeightConstraint = [self.tableView.heightAnchor constraintEqualToConstant:100];
    [constraints addObjectsFromArray:@[[self.tableView.topAnchor constraintEqualToAnchor:topTableVieWConstraints constant:verticalPadding],
                                       [self.tableView.leadingAnchor constraintEqualToAnchor:self.contentScrollView.leadingAnchor constant:20],
                                       [self.tableView.trailingAnchor constraintEqualToAnchor:self.contentScrollView.trailingAnchor constant:-20],
                                       self.tableViewHeightConstraint,
                                       [self.tableView.bottomAnchor constraintEqualToAnchor:self.firstImageView.topAnchor constant:-verticalPadding]]];
    
    CGFloat ratio = self.documentImage.size.height / self.documentImage.size.width;
    CGFloat width = self.view.bounds.size.width-(horizontalPadding*2);
    CGFloat height = width*ratio;
    [constraints addObjectsFromArray:@[[self.firstImageView.topAnchor constraintEqualToAnchor:self.tableView.bottomAnchor constant:verticalPadding],
                                       [self.firstImageView.leadingAnchor constraintEqualToAnchor:self.contentScrollView.leadingAnchor constant:horizontalPadding],
                                       [self.firstImageView.trailingAnchor constraintEqualToAnchor:self.contentScrollView.trailingAnchor constant:-horizontalPadding],
                                       [self.firstImageView.heightAnchor constraintLessThanOrEqualToConstant:isnan(height) ? 0 : height],
                                       [self.firstImageView.widthAnchor constraintEqualToConstant:isnan(width) ? 0 : width],
                                       [self.firstImageView.bottomAnchor constraintEqualToAnchor:self.secondImageView.topAnchor constant:-verticalPadding]]];

    ratio = self.documentBackImage.size.height / self.documentBackImage.size.width;
    height = width*ratio;
    NSLayoutAnchor *secondImageViewBottomAnchor = _shouldShowDisclaimer ? self.disclaimerTextView.topAnchor : self.contentScrollView.bottomAnchor;
    [constraints addObjectsFromArray:@[[self.secondImageView.topAnchor constraintEqualToAnchor:self.firstImageView.bottomAnchor constant:verticalPadding],
                                       [self.secondImageView.leadingAnchor constraintEqualToAnchor:self.contentScrollView.leadingAnchor constant:horizontalPadding],
                                       [self.secondImageView.trailingAnchor constraintEqualToAnchor:self.contentScrollView.trailingAnchor constant:-horizontalPadding],
                                       [self.secondImageView.heightAnchor constraintEqualToConstant:isnan(height) ? 0 : height],
                                       [self.secondImageView.widthAnchor constraintEqualToAnchor:self.firstImageView.widthAnchor],
                                       [self.secondImageView.bottomAnchor constraintEqualToAnchor:secondImageViewBottomAnchor constant:-verticalPadding]]];
    if (_shouldShowDisclaimer) {
        CGSize size = [self.disclaimerTextView sizeThatFits:CGSizeMake(width, 100)];
        CGFloat disclaimerHeight = size.height;
        [constraints addObjectsFromArray:@[[self.disclaimerTextView.topAnchor constraintEqualToAnchor:self.secondImageView.bottomAnchor constant:verticalPadding],
                                           [self.disclaimerTextView.leadingAnchor constraintEqualToAnchor:self.contentScrollView.leadingAnchor constant:horizontalPadding],
                                           [self.disclaimerTextView.trailingAnchor constraintEqualToAnchor:self.contentScrollView.trailingAnchor constant:-horizontalPadding],
                                           [self.disclaimerTextView.bottomAnchor constraintEqualToAnchor:self.contentScrollView.bottomAnchor constant:-verticalPadding],
                                           [self.disclaimerTextView.heightAnchor constraintEqualToConstant:disclaimerHeight]]];
    }
    
    [constraints addObjectsFromArray:@[[self.confirmButton.topAnchor constraintEqualToAnchor:self.contentScrollView.bottomAnchor constant:verticalPadding],
                                       [self.confirmButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-horizontalPadding],
                                       [self.confirmButton.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:horizontalPadding],
                                       [self.confirmButton.heightAnchor constraintEqualToConstant:50],
                                       [self.confirmButton.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-verticalPadding]]];
    
    
    [self.view addConstraints:constraints];
    [NSLayoutConstraint activateConstraints:constraints];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.navigationController.toolbarHidden = YES;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    [self.tableView layoutIfNeeded];
    self.tableViewHeightConstraint.constant = self.tableView.contentSize.height;
}

#pragma mark - UITableView methods

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"alResultCell";
    ALResultCell *cell;
    cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[ALResultCell alloc] init];
    }
    
    ALResultEntry *entry = [self.resultData[[[self.resultData allKeys] objectAtIndex:indexPath.section]] objectAtIndex:indexPath.row];
    cell.backgroundColor = [UIColor AL_BackgroundColor];
    [cell setResultEntry:entry];
    if (self.isArabicScript) {
        [cell alignLabelsTextRight];
    }
    
    
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSInteger rowCount = [self.resultData[[[self.resultData allKeys] objectAtIndex:section]] count];
    return self.isArabicScript ? rowCount - 1 : rowCount;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    UIWindow *window = UIApplication.sharedApplication.keyWindow;
    CGFloat topPadding = window.safeAreaInsets.top;
    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(0, topPadding, self.view.frame.size.width, [self headerSize])];
    view.backgroundColor = [UIColor AL_BackgroundColor];
    
    return view;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [self.resultData allKeys].count;
}

#pragma mark - Utility methods

- (CGFloat)headerSize {
    return 7.0;
}

- (void)setHyperlinkInsideTextView {
    NSMutableAttributedString *attString = [[NSMutableAttributedString alloc] initWithString:disclaimerString];
    NSRange linkRange = [[attString mutableString] rangeOfString:@"documentation"];
    [attString addAttributes:@{NSLinkAttributeName : [NSURL URLWithString:WEBLINK_ANYLINE_DOCUMENTATION_PRODUCTID]} range:linkRange];
    
    [self.disclaimerTextView setAttributedText:[attString withFont:[UIFont AL_proximaRegularWithSize:16]]];
    [self.disclaimerTextView setTextColor:[UIColor AL_LabelBlackWhite]];
}

#pragma mark - IBAction methods
- (void)confirmAction:(id)sender {
    [self.navigationController popViewControllerAnimated:YES];
}

- (BOOL)tableView:(UITableView*)tableView canPerformAction:(SEL)action forRowAtIndexPath:(NSIndexPath*)indexPath withSender:(id)sender {
    
    if (action == @selector(copy:)) {
        return YES;
    }
    
    return NO;
}

- (BOOL)tableView:(UITableView*)tableView shouldShowMenuForRowAtIndexPath:(NSIndexPath*)indexPath {
    return YES;
}

-(void)tableView:(UITableView*)tableView performAction:(SEL)action forRowAtIndexPath:(NSIndexPath*)indexPath withSender:(id)sender {
    
    UIPasteboard* pasteboard = [UIPasteboard generalPasteboard];
    
    ALResultEntry *entry = [self.resultData[[[self.resultData allKeys] objectAtIndex:indexPath.section]] objectAtIndex:indexPath.row];
    
    pasteboard.string = entry.value;
}

@end

File: AnylineExamples/Anyline Examples Source/Utility/View/ResultScreen/ALResultViewController.swift
//
//  ALResultViewController.swift
//  AnylineExamples
//
//  Created by Aldrich Co on 3/11/22.
//

import Foundation

class ALResultViewController: UIViewController {

    enum Constants {
        static let disclaimerStr = """
The result fields above display a selection of scannable ID information only. Please review the documentation for a full list of scannable ID information.
"""
        static let idPluginDocumentationURL = "https://documentation.anyline.com/toc/products/id/index.html"

        static let keysToExclude = ["isArabic"]

        static let defaultSectionTitle = "Result Data"
    }

    // the keys can be used as titles for section headers, we choose not to display them
    var resultData: [String: [ALResultEntry]] = [:]

    @objc
    var imageFace: UIImage?

    @objc
    var imagePrimary: UIImage?

    @objc
    var imageSecondary: UIImage?

    @objc
    var showDisclaimer = false

    @objc
    var isRightToLeft = false

    private var tableViewHeightConstraint: NSLayoutConstraint!

    lazy var contentScrollView: UIScrollView = {
        let scrollView = UIScrollView()
        scrollView.backgroundColor = .al_Background()
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        return scrollView
    }()

    lazy var faceImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        imageView.translatesAutoresizingMaskIntoConstraints = false
        return imageView
    }()

    lazy var firstImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        imageView.translatesAutoresizingMaskIntoConstraints = false
        return imageView
    }()

    lazy var secondImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        imageView.translatesAutoresizingMaskIntoConstraints = false
        return imageView
    }()

    lazy var confirmButton: UIButton = {
        let button = UIButton()
        button.addTarget(self, action: #selector(confirmAction(_:)), for: .touchUpInside)
        button.setTitle("Continue scanning", for: .normal)
        button.titleLabel?.font = .al_proximaBold(withSize: 18)
        button.setTitleColor(.white, for: .normal)
        button.backgroundColor = .al_examplesBlue()
        button.layer.cornerRadius = 25
        button.translatesAutoresizingMaskIntoConstraints = false
        return button
    }()

    lazy var disclaimerTextView: UITextView = {
        let textView = UITextView()
        textView.isScrollEnabled = false
        textView.clipsToBounds = false
        textView.isUserInteractionEnabled = true
        textView.isEditable = false
        textView.translatesAutoresizingMaskIntoConstraints = false
        return textView
    }()

    lazy var tableView: UITableView = {
        let tableView = UITableView()
        tableView.register(ALResultCell.self,
                           forCellReuseIdentifier: "cell")
        tableView.isScrollEnabled = false
        tableView.allowsSelection = false
        tableView.backgroundColor = .al_Background()
        tableView.separatorStyle = .none
        tableView.estimatedRowHeight = 60
        tableView.estimatedSectionHeaderHeight = 0
        tableView.estimatedSectionFooterHeight = 0
        tableView.rowHeight = UITableView.automaticDimension
        tableView.tableFooterView = UIView()
        tableView.translatesAutoresizingMaskIntoConstraints = false
        return tableView
    }()

    @objc
    func confirmAction(_ button: UIButton) {
        self.navigationController?.popViewController(animated: true)
    }

    @objc
    init(resultMap: [String: [ALResultEntry]]) {
        self.resultData = resultMap.excludingKeys(Constants.keysToExclude)
        super.init(nibName: nil, bundle: nil)
    }

    @objc
    convenience init(results: [ALResultEntry]) {
        self.init(resultMap: [ Constants.defaultSectionTitle: results ])
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        layout()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setToolbarHidden(true,
                                                    animated: false)
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        self.updateLayout()
    }

    private func layout() {

        self.setupNavBar()

        self.view.backgroundColor = .al_Background()
        self.contentScrollView.delegate = self

        self.view.addSubview(self.contentScrollView)
        self.view.addSubview(self.confirmButton)

        self.faceImageView.image = self.imageFace
        self.firstImageView.image = self.imagePrimary
        self.secondImageView.image = self.imageSecondary

        contentScrollView.addSubview(self.faceImageView)
        contentScrollView.addSubview(self.firstImageView)
        contentScrollView.addSubview(self.secondImageView)

        self.tableView.delegate = self
        self.tableView.dataSource = self

        contentScrollView.addSubview(self.tableView)

        if showDisclaimer {
            contentScrollView.addSubview(self.disclaimerTextView)
            self.setDisclaimerTextViewContent()
        }

        setupConstraints()
        updateLayout()
    }

    private func setupNavBar() {
        self.title = "Result Data"
        let shareBarItem: UIBarButtonItem = .init(barButtonSystemItem: .action,
                                                  target: self,
                                                  action: #selector(askToShareItems(_:)))
        shareBarItem.tintColor = .al_LabelBlackWhite()
        self.navigationItem.rightBarButtonItem = shareBarItem
    }

    private func setupConstraints() {

        let verticalGap: CGFloat = 10.0
        let horizontalGap: CGFloat = 15.0

        var allConstraints: [NSLayoutConstraint] = [
            self.contentScrollView.topAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.topAnchor),
            self.contentScrollView.leadingAnchor.constraint(equalTo: self.view.leadingAnchor),
            self.contentScrollView.trailingAnchor.constraint(equalTo: self.view.trailingAnchor),
        ]

        var tableViewTopConstraint: NSLayoutYAxisAnchor = self.contentScrollView.topAnchor

        if self.imageFace != nil {
            var horizontalConstraint: NSLayoutConstraint = self.faceImageView.leadingAnchor
                .constraint(equalTo: self.tableView.leadingAnchor,
                            constant: horizontalGap - 5)
            if self.isRightToLeft {
                horizontalConstraint = self.faceImageView.trailingAnchor
                    .constraint(equalTo: self.tableView.trailingAnchor,
                                constant: -horizontalGap + 5)
            }

            allConstraints.append(contentsOf: [
                self.faceImageView.topAnchor.constraint(equalTo: self.contentScrollView.topAnchor,
                                                        constant: verticalGap),
                horizontalConstraint,
                self.faceImageView.widthAnchor.constraint(lessThanOrEqualToConstant: 100),
                self.faceImageView.heightAnchor.constraint(lessThanOrEqualToConstant: 100),
            ])

            tableViewTopConstraint = self.faceImageView.bottomAnchor
        }

        // 100 is arbitrary, it will be updated to the fitting height after a runloop update
        self.tableViewHeightConstraint = self.tableView.heightAnchor.constraint(equalToConstant: 100)
        allConstraints.append(contentsOf: [
            self.tableView.topAnchor.constraint(equalTo: tableViewTopConstraint, constant: verticalGap),
            self.tableView.leadingAnchor.constraint(equalTo: self.view.leadingAnchor, constant: horizontalGap),
            self.tableView.trailingAnchor.constraint(equalTo: self.view.trailingAnchor, constant: -horizontalGap),
            self.tableViewHeightConstraint
        ])

        var ratio = (self.imagePrimary?.size.height ?? 0) / (self.imagePrimary?.size.width ?? 0)
        let width = self.view.bounds.width - horizontalGap * 2.0
        var height = width * ratio

        allConstraints.append(contentsOf: [
            self.firstImageView.topAnchor.constraint(equalTo: self.tableView.bottomAnchor, constant: verticalGap),
            self.firstImageView.leadingAnchor.constraint(equalTo: self.tableView.leadingAnchor, constant: 10),
            self.firstImageView.trailingAnchor.constraint(equalTo:self.tableView.trailingAnchor, constant: -10),
            self.firstImageView.heightAnchor.constraint(lessThanOrEqualToConstant: height.isNaN ? 0 : height),
            self.firstImageView.widthAnchor.constraint(lessThanOrEqualToConstant: width.isNaN ? 0 : width),
            self.firstImageView.bottomAnchor.constraint(equalTo: self.secondImageView.topAnchor, constant: -verticalGap)
        ])
        ratio = (self.imageSecondary?.size.height ?? 0) / (self.imageSecondary?.size.width ?? 0)
        height = width * ratio

        let secondImageViewBottomAnchor = (self.showDisclaimer ?
                                           self.disclaimerTextView.topAnchor :
                                            self.contentScrollView.bottomAnchor)

        allConstraints.append(contentsOf: [
            self.secondImageView.topAnchor.constraint(equalTo: self.firstImageView.bottomAnchor, constant: verticalGap),
            self.secondImageView.leadingAnchor.constraint(equalTo: self.tableView.leadingAnchor, constant: 10),
            self.secondImageView.trailingAnchor.constraint(equalTo: self.tableView.trailingAnchor, constant: -10),
            self.secondImageView.heightAnchor.constraint(equalToConstant: height.isNaN ? 0 : height),
            self.secondImageView.widthAnchor.constraint(equalTo: self.firstImageView.widthAnchor),
            self.secondImageView.bottomAnchor.constraint(equalTo: secondImageViewBottomAnchor, constant: -verticalGap)
        ])

        if self.showDisclaimer {
            let size = self.disclaimerTextView.sizeThatFits(.init(width: width, height: 100))
            let disclaimerHeight = size.height

            allConstraints.append(contentsOf: [
                self.disclaimerTextView.topAnchor.constraint(equalTo: self.secondImageView.bottomAnchor, constant: verticalGap),
                self.disclaimerTextView.leadingAnchor.constraint(equalTo: self.tableView.leadingAnchor, constant: 10),
                self.disclaimerTextView.trailingAnchor.constraint(equalTo: self.tableView.trailingAnchor, constant: -10),
                self.disclaimerTextView.bottomAnchor.constraint(equalTo: self.contentScrollView.bottomAnchor, constant: -verticalGap),
                self.disclaimerTextView.heightAnchor.constraint(equalToConstant: disclaimerHeight)
            ])
        }

        allConstraints.append(contentsOf: [
            self.confirmButton.topAnchor.constraint(equalTo: self.contentScrollView.bottomAnchor,
                                                    constant: verticalGap),
            self.confirmButton.leadingAnchor.constraint(equalTo: self.tableView.leadingAnchor),
            self.confirmButton.trailingAnchor.constraint(equalTo: self.tableView.trailingAnchor),
            self.confirmButton.heightAnchor.constraint(equalToConstant: 50),
            self.confirmButton.bottomAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.bottomAnchor,
                                                       constant: -verticalGap * 2)
        ])

        self.view.addConstraints(allConstraints)
        NSLayoutConstraint.activate(allConstraints)
    }

    private func updateLayout() {
        self.tableView.layoutIfNeeded()
        self.tableViewHeightConstraint.constant = self.tableView.contentSize.height
    }

    private func setDisclaimerTextViewContent() {
        // sets the disclaimer text found at the bottom of the page, should you
        // wish to allow it to be shown (typically this is done on ID scans)
        let attrStr: NSMutableAttributedString = .init(string: Constants.disclaimerStr)
        let hyperlinkRange = attrStr.mutableString.range(of: "documentation")
        let idPluginDocumentationURL = URL(string: Constants.idPluginDocumentationURL)!
        attrStr.addAttribute(.link, value: idPluginDocumentationURL, range: hyperlinkRange)
        self.disclaimerTextView.attributedText = attrStr
        self.disclaimerTextView.textColor = .al_LabelBlackWhite()
    }

    @objc func askToShareItems(_ item: Any) {
        let resultArr = self.resultData["Result Data"] ?? []
        var shareText: String = "Anyline Scanner Result\n\n"
        for entry in resultArr {
            shareText.append(String(format: "%@: %@\n", entry.title, entry.value))
        }
        let items = [ shareText, self.imagePrimary as Any ].compactMap { $0 }
        let activityController = UIActivityViewController(activityItems: items,
                                                          applicationActivities: nil)
        activityController.setValue("Anyline Scanner Result", forKey: "Subject")
        if UIDevice().userInterfaceIdiom != .phone {
            activityController.modalPresentationStyle = .popover
        }
        self.present(activityController, animated: true, completion: nil)
    }
}

extension ALResultViewController: UITableViewDelegate, UITableViewDataSource, UIScrollViewDelegate {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let keys = Array<String>(self.resultData.keys)
        return self.resultData[keys[section]]?.count ?? 0
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let resultCell = tableView.dequeueReusableCell(withIdentifier: "cell") as? ALResultCell ?? ALResultCell()
        let keys = Array<String>(self.resultData.keys)
        let key = keys[indexPath.section]
        if let section = self.resultData[key] {
            let entry = section[indexPath.row]
            resultCell.backgroundColor = .al_Background()
            resultCell.resultEntry = entry
            resultCell.alignLabels(.left)
            if self.isRightToLeft {
                resultCell.alignLabels(.right)
            }
            return resultCell
        }
        return UITableViewCell()
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return self.resultData.keys.count
    }

    func tableView(_ tableView: UITableView,
                   shouldShowMenuForRowAt indexPath: IndexPath) -> Bool {
        return true
    }

    func tableView(_ tableView: UITableView, canPerformAction action: Selector,
                   forRowAt indexPath: IndexPath, withSender sender: Any?) -> Bool {
        if action == #selector(UIResponderStandardEditActions.copy(_:)) {
            return true
        }
        return false
    }

    func tableView(_ tableView: UITableView, performAction action: Selector,
                   forRowAt indexPath: IndexPath, withSender sender: Any?) {
        let pasteboard = UIPasteboard.general
        let keys = Array<String>(self.resultData.keys)
        let section = self.resultData[keys[indexPath.section]]
        if let entry = section?[indexPath.row] {
            pasteboard.string = entry.value
        }
    }
}


extension Dictionary where Key == String, Value == Array<ALResultEntry> {

    /// This is run over the entries to be passed as input to this view controller,
    /// to remove excluded elements such as `isArabic`.
    ///
    /// - Parameter excludedKeys: array of keys to exclude
    /// - Returns: the dictionary with the given keys excluded / removed
    fileprivate func excludingKeys(_ excludedKeys: [String]) -> Dictionary<Key, Value> {
        var result = [String: [ALResultEntry]]()
        for k in self.keys {
            if let arr: [ALResultEntry] = self[k] {
                result[k] = arr.filter { entry in
                    let entryKey = entry.title ?? ""
                    return !excludedKeys.contains(entryKey)
                }
            }
        }
        return result
    }
}
